rfcs/20180905-deprecate-collections.md

```diff
@@ -0,0 +1,575 @@
+
+The concern with this solution is that in serving, the first request would take more time than the rest of the calls and that sort of inconsistent performance is not desirable. 
+
+
+### **Asset file paths**
+
+Asset files currently only include vocabulary files that are used to initialize tables with. The tracking of tables should be sufficient to track vocabulary files. 
+
+
+### **Summaries**
+
+Summary V1 uses collection extensively but the new Summary V2 implementation uses collections minimally. Those uses of collections all fall under Case 3 and are replaceable with global variables. Summary V1 deprecation is part of TF 2.0 plans as of now.
+
+
+### **Iterators**
+
+(`GLOBAL_ITERATORS` collection) Iterators are checkpointable (saveable) objects and we should handle them just as we handle other non-variable saveable objects. The challenge here is that iterators are objects that are created outside the model function and in fact by the [estimator code itself](https://github.com/tensorflow/tensorflow/blob/r1.10/tensorflow/python/estimator/estimator.py#L1035). As a result, in the regular object based checkpointing flow, no "user" object would be depending on the iterator. Hence, before passing the saveables to the Saver, we'll staple together the iterator created by the estimator with the saveables from the user created object. We might want to control this behaviour with a RunConfig flag somewhere perhaps.
+
+
+### **Queue runners**
+
+Queue runners are used by two use cases that should go away with TF 2.0
+
+1.  **Input pipelines:** tf.data is the replacement and it doesn't use queues at all
+1.  **Sync Replicas Optimizer:** Distribution strategies is the replacement
+
+
+### **Conditionals**
+
+The current Cond implementation uses collections but the new CondV2 implementation doesn't rely on it. Although moving to CondV2 is not part of TF 2.0 plans since its not a breaking change but it's quite likely that the timelines would line up. This effort (deprecating collections) would introduce a dependency on that effort.
+
+
+### **Updates**
+
+Batch normalization is the only use case for updates. For keras models, all updates are tracked via Model.updates. For the graph construction case, the updates are accessible via the updates attribute on the BatchNormalization layer itself and then they could be added to the train op. We will deprecate the [functional batch_normalization API](https://github.com/tensorflow/tensorflow/blob/r1.10/tensorflow/python/layers/normalization.py#L158) so that users deal with objects only.
+
```

Comment:
Can you provide an example of what this would look like for user code?

Reply:
Done.